File: golang/3203_maximum_length/maximum_length.go
// Package main ...
package main

import (
	"testing"

	"github.com/stretchr/testify/assert"
)

func maximumLength(nums []int, k int) int {
	dp := make([][]int, k)
	for i := range dp {
		dp[i] = make([]int, k)
	}
	res := 0
	for _, num := range nums {
		num %= k
		for prev := 0; prev < k; prev++ {
			dp[prev][num] = dp[num][prev] + 1
			res = max(res, dp[prev][num])
		}
	}
	return res
}

func main() {
	tests := []struct {
		nums []int
		k    int
		ans  int
	}{
		{[]int{1, 2, 3, 4, 5}, 2, 5},
		{[]int{1, 4, 2, 3, 1, 4}, 3, 4},
	}

	for index, test := range tests {
		assert.Equal(&testing.T{}, test.ans, maximumLength(test.nums, test.k), "case %d", index)
	}
}

File: cpp/3203_maximum_length/maximum_length.cpp
#include <cassert>
#include <tuple>
#include <vector>

using namespace std;

class Solution {
 public:
  int maximumLength(vector<int>& nums, int k) {
    vector<vector<int>> dp(k, vector<int>(k, 0));
    int res = 0;
    for (int num : nums) {
      num %= k;
      for (int prev = 0; prev < k; ++prev) {
        dp[prev][num] = dp[num][prev] + 1;
        res = max(res, dp[prev][num]);
      }
    }
    return res;
  }
};

int main() {
  vector<tuple<vector<int>, int, int>> tests{
      {{1, 2, 3, 4, 5}, 2, 5},
      {{1, 4, 2, 3, 1, 4}, 3, 4},
  };

  for (auto& [nums, k, expected] : tests) {
    assert(Solution().maximumLength(nums, k) == expected);
  }
}